File: content/problems/1900-the-earliest-and-latest-rounds-where-players-compete/solutions/round-state-outcome-dfs/solution.cpp
class Solution {
    map<tuple<int,int,int>,array<int,2>> memo;
    void enumerate(int n,int first,int second,int pair,int beforeFirst,int beforeSecond,set<int>& states){
        if(pair>n/2){
            if(n%2){
                int middle=(n+1)/2;
                beforeFirst+=middle<first;
                beforeSecond+=middle<second;
            }
            states.insert(beforeFirst*32+beforeSecond);
            return;
        }
        int a=pair,b=n+1-pair;
        auto advance=[&](int winner){
            enumerate(n,first,second,pair+1,beforeFirst+(winner<first),beforeSecond+(winner<second),states);
        };
        if(a==first||b==first)advance(first);
        else if(a==second||b==second)advance(second);
        else{
            advance(a);
            advance(b);
        }
    }
    array<int,2> solve(int n,int first,int second){
        if(first+second==n+1)return {
            1,1
        };
        if(first+second>n+1){
            int old=first;
            first=n+1-second;
            second=n+1-old;
        }
        auto key=make_tuple(n,first,second);
        if(memo.count(key))return memo[key];
        set<int> states;
        enumerate(n,first,second,1,0,0,states);
        int earliest=100,latest=0,nextN=(n+1)/2;
        for(int state:states){
            auto result=solve(nextN,state/32+1,state%32+1);
            earliest=min(earliest,result[0]+1);
            latest=max(latest,result[1]+1);
        }
        return memo[key]={
            earliest,latest
        };
    }
    public: vector<int> earliestAndLatest(int n,int firstPlayer,int secondPlayer){
        auto result=solve(n,firstPlayer,secondPlayer);
        return {
            result[0],result[1]
        };
    }
};

File: content/problems/1900-the-earliest-and-latest-rounds-where-players-compete/solutions/round-state-outcome-dfs/Solution.kt
class Solution {
    private val memo=HashMap<String,IntArray>();
    private fun enumerate(n:Int,first:Int,second:Int,pair:Int,beforeFirst:Int,beforeSecond:Int,states:MutableSet<Int>){
        if(pair>n/2){
            var a=beforeFirst;
            var b=beforeSecond;
            if(n%2==1){
                val middle=(n+1)/2;
                if(middle<first)a++;
                if(middle<second)b++
            };
            states.add(a*32+b);
            return
        };
        val left=pair;
        val right=n+1-pair;
        fun advance(winner:Int)=enumerate(n,first,second,pair+1,beforeFirst+if(winner<first)1 else 0,beforeSecond+if(winner<second)1 else 0,states);
        when{
            left==first||right==first->advance(first);
            left==second||right==second->advance(second);
            else->{
                advance(left);
                advance(right)
            }
        }
    };
    private fun solve(n:Int,firstValue:Int,secondValue:Int):IntArray{
        if(firstValue+secondValue==n+1)return intArrayOf(1,1);
        var first=firstValue;
        var second=secondValue;
        if(first+second>n+1){
            val old=first;
            first=n+1-second;
            second=n+1-old
        };
        val key="$n,$first,$second";
        memo[key]?.let{
            return it
        };
        val states=HashSet<Int>();
        enumerate(n,first,second,1,0,0,states);
        var earliest=100;
        var latest=0;
        val nextN=(n+1)/2;
        for(state in states){
            val result=solve(nextN,state/32+1,state%32+1);
            earliest=minOf(earliest,result[0]+1);
            latest=maxOf(latest,result[1]+1)
        };
        return intArrayOf(earliest,latest).also{
            memo[key]=it
        }
    };
    fun earliestAndLatest(n:Int,firstPlayer:Int,secondPlayer:Int):IntArray=solve(n,firstPlayer,secondPlayer)
}
